**AUTHORS DETAILS:**

Athul Narayanan(249426510)

Anitha Thomas(249417370)
    
Parth Sathiya(259662610)
    
Varna SalimKumar(259694440)

**Project Abstract:**

Our project focuses on analyzing US Senate election results from multiple years (1976-2020) to find patterns and trends in the election. The overall goal of the project is not only to analyze the election results but also find political dominance of parties across different states and how they are deviating during the period. It will also contribute to discussions about future electoral patterns and political forecasting in addition to offering an overview of previous elections through data-driven insights and visual narratives.

**Dataset Resource and information:**

https://dataverse.harvard.edu/file.xhtml?fileId=7609736&version=7.0&toolType=PREVIEW



In [ ]:
import numpy as np
import pandas as pd
import seaborn as sns
import plotly.express as px
import matplotlib.pyplot as plt


In [ ]:
df = pd.read_csv('./1976-2020-senate.csv',usecols=['year', 'state', 'state_po', 'state_fips', 'state_cen', 'state_ic',
       'office', 'district', 'stage', 'special', 'candidate', 'party_detailed',
       'writein', 'mode', 'candidatevotes', 'totalvotes', 'unofficial',
       'version', 'party_simplified'])
data_copy = df.copy(deep=True)

In [ ]:
len(df.columns)

19

In [ ]:
# Need to load file named 'US_States_info.xlsx'
US_state_info  = pd.read_excel('./US_States_info.xlsx',usecols=['STATE', 'STATE_PO', 'STATE_FIPS','STATE_CENSUS_CODE', 'STATE_ICPSR'])

In [ ]:
US_state_info['STATE'] = US_state_info['STATE'].str.upper()

In [ ]:
df.head()

,year,state,state_po,state_fips,state_cen,state_ic,office,district,stage,special,candidate,party_detailed,writein,mode,candidatevotes,totalvotes,unofficial,version,party_simplified
0,1976,ARIZONA,AZ,4,86,61,US SENATE,statewide,gen,False,SAM STEIGER,REPUBLICAN,False,total,321236,741210,False,20210114,REPUBLICAN
1,1976,ARIZONA,AZ,4,86,61,US SENATE,statewide,gen,False,WM. MATHEWS FEIGHAN,INDEPENDENT,False,total,1565,741210,False,20210114,OTHER
2,1976,ARIZONA,AZ,4,86,61,US SENATE,statewide,gen,False,DENNIS DECONCINI,DEMOCRAT,False,total,400334,741210,False,20210114,DEMOCRAT
3,1976,ARIZONA,AZ,4,86,61,US SENATE,statewide,gen,False,ALLAN NORWITZ,LIBERTARIAN,False,total,7310,741210,False,20210114,LIBERTARIAN
4,1976,ARIZONA,AZ,4,86,61,US SENATE,statewide,gen,False,BOB FIELD,INDEPENDENT,False,total,10765,741210,False,20210114,OTHER


In [ ]:
df

,year,state,state_po,state_fips,state_cen,state_ic,office,district,stage,special,candidate,party_detailed,writein,mode,candidatevotes,totalvotes,unofficial,version,party_simplified
0,1976,ARIZONA,AZ,4,86,61,US SENATE,statewide,gen,False,SAM STEIGER,REPUBLICAN,False,total,321236,741210,False,20210114,REPUBLICAN
1,1976,ARIZONA,AZ,4,86,61,US SENATE,statewide,gen,False,WM. MATHEWS FEIGHAN,INDEPENDENT,False,total,1565,741210,False,20210114,OTHER
2,1976,ARIZONA,AZ,4,86,61,US SENATE,statewide,gen,False,DENNIS DECONCINI,DEMOCRAT,False,total,400334,741210,False,20210114,DEMOCRAT
3,1976,ARIZONA,AZ,4,86,61,US SENATE,statewide,gen,False,ALLAN NORWITZ,LIBERTARIAN,False,total,7310,741210,False,20210114,LIBERTARIAN
4,1976,ARIZONA,AZ,4,86,61,US SENATE,statewide,gen,False,BOB FIELD,INDEPENDENT,False,total,10765,741210,False,20210114,OTHER
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
3624,2020,WYOMING,WY,56,83,68,US SENATE,statewide,gen,False,UNDER VOTES,NaN,False,total,6401,278503,False,20210114,OTHER
3625,2021,GEORGIA,GA,13,58,44,US SENATE,statewide,runoff,True,KELLY LOEFFLER,REPUBLICAN,False,total,2194848,4483294,True,20210114,REPUBLICAN
3626,2021,GEORGIA,GA,13,58,44,US SENATE,statewide,runoff,True,RAPHAEL WARNOCK,DEMOCRAT,False,total,2288446,4483294,True,20210114,DEMOCRAT
3627,2021,GEORGIA,GA,13,58,44,US SENATE,statewide,runoff,False,DAVID A. PERDUE,REPUBLICAN,False,total,2213979,4483241,True,20210114,REPUBLICAN


# Data Preprocessing

In [ ]:
df.shape

(3629, 19)

In [ ]:
df.columns

Index(['year', 'state', 'state_po', 'state_fips', 'state_cen', 'state_ic',
       'office', 'district', 'stage', 'special', 'candidate', 'party_detailed',
       'writein', 'mode', 'candidatevotes', 'totalvotes', 'unofficial',
       'version', 'party_simplified'],
      dtype='object')

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3629 entries, 0 to 3628
Data columns (total 19 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   year              3629 non-null   int64 
 1   state             3629 non-null   object
 2   state_po          3629 non-null   object
 3   state_fips        3629 non-null   int64 
 4   state_cen         3629 non-null   int64 
 5   state_ic          3629 non-null   int64 
 6   office            3629 non-null   object
 7   district          3629 non-null   object
 8   stage             3629 non-null   object
 9   special           3629 non-null   bool  
 10  candidate         3206 non-null   object
 11  party_detailed    3030 non-null   object
 12  writein           3629 non-null   bool  
 13  mode              3629 non-null   object
 14  candidatevotes    3629 non-null   int64 
 15  totalvotes        3629 non-null   int64 
 16  unofficial        3629 non-null   bool  
 17  version       

In [ ]:
df.isnull().sum()


,0
year,0
state,0
state_po,0
state_fips,0
state_cen,0
state_ic,0
office,0
district,0
stage,0
special,0


#ATTRIBUTE DESCRIPTION

**YEAR**: The year of the election (1976 in this dataset).

**STATE**: The full name of the state where the election took place (ARIZONA).

**STATE_PO**: The postal abbreviation of the state (AZ for Arizona).

**STATE_FIPS**: The Federal Information Processing Standard (FIPS) code for the state (4 for Arizona).
FIPS codes are unique numerical identifiers assigned to geographic areas.

**STATE_CEN**: Census Code for the state (86 for Arizona).
Used by the U.S. Census Bureau for statistical purposes.

**STATE_IC**: Internal Census Bureau code for the state (61 for Arizona).
Similar to state_cen, it’s an internal identifier.

**OFFICE**: The political office contested in the election (US SENATE in this dataset).
Other values in a broader dataset might include "US HOUSE," "GOVERNOR," etc.

**DISTRICT**: The congressional district for the election.
Since U.S. Senate races are state-wide, this value is always "statewide".

**STAGE**: The stage of the election, here "gen" (General Election).
Possible values in a broader dataset:
"gen" → General election
"pri" → Primary election
"runoff" → Runoff election

**SPECIAL** Indicates if the election was a special election (due to vacancy, etc.).
False means it was a regular election.

**CANDIDATE**: The name of the candidate running in the election.

**PARTY_DETAILED**: The detailed political party affiliation of the candidate, e.g., "REPUBLICAN", "DEMOCRAT", "LIBERTARIAN", "INDEPENDENT"

**WRITEIN**: True if the candidate was a write-in (not officially on the ballot).
False means they were officially listed on the ballot.

**MODE**: The method of voting recorded, always "total" in this dataset.
Could indicate specific voting methods (mail-in, early voting, etc.) in broader datasets.

**CANDIDATEVOTES**: The total number of votes received by the candidate.

**TOTALVOTES**: The total number of votes cast in the election for all candidates combined (741,210 in this case).

**UNOFFICIAL**: Indicates whether the election result is official or unofficial.
False means the results are official.

**VERSION**: A timestamp or version number for data tracking purposes.
"20210114" suggests the dataset was last updated on January 14, 2021.

**PARTY_SIMPLIFIED**: A simplified version of party_detailed, categorizing candidates into broader
political groups:
"REPUBLICAN" → Republican Party
"DEMOCRAT" → Democratic Party
"OTHER" → Independents or minor parties
"LIBERTARIAN" → Libertarian Party



#DATA VALIDATION

**YEAR:** The year of the election.


*   Value should be within the range of 1976 to 2020



In [ ]:
df[(df['year'] < 1976) | (df['year'] > 2020)]
# Removing this records

,year,state,state_po,state_fips,state_cen,state_ic,office,district,stage,special,candidate,party_detailed,writein,mode,candidatevotes,totalvotes,unofficial,version,party_simplified
3625,2021,GEORGIA,GA,13,58,44,US SENATE,statewide,runoff,True,KELLY LOEFFLER,REPUBLICAN,False,total,2194848,4483294,True,20210114,REPUBLICAN
3626,2021,GEORGIA,GA,13,58,44,US SENATE,statewide,runoff,True,RAPHAEL WARNOCK,DEMOCRAT,False,total,2288446,4483294,True,20210114,DEMOCRAT
3627,2021,GEORGIA,GA,13,58,44,US SENATE,statewide,runoff,False,DAVID A. PERDUE,REPUBLICAN,False,total,2213979,4483241,True,20210114,REPUBLICAN
3628,2021,GEORGIA,GA,13,58,44,US SENATE,statewide,runoff,False,JON OSSOFF,DEMOCRAT,False,total,2269262,4483241,True,20210114,DEMOCRAT


In [ ]:
# Filtering out the records which falls out of defined range which is 1976 to 2020
df = df[(df['year'] >= 1976) & (df['year'] <= 2020)]

In [ ]:
fig = px.histogram(df, x="year", title="Distribution of Years", color="year", nbins=len(df['year'].unique()))

fig.update_layout(bargap=0.2)

fig.show()

**STATE**: The full name of the state where the election took place.



*   All the values should be validate US States Name.
*   Need to create list for valid states of US that exist during 1976-2020 by studing external resources.


- @Anitha & @Verna
- Note: if possible please add the reference of the resources





In [ ]:
valid_state_lst = []
valid_state_lst = US_state_info['STATE'].str.upper().tolist()

In [ ]:
isValid_state_lst = df['state'].isin(valid_state_lst)

In [ ]:
if isValid_state_lst.all():
  print('All the values under state columns are validated')
else:
  print('All the values under state columns are not validated')
  print('Invalid Values are')
  print(df[~isValid_state_lst])

All the values under state columns are validated


**STATE_PO**: The postal abbreviation of the state



*   All the values should be valid postal abbreviation.
*   Need to create list for valid  postal abbreviation for each states of US that exist during 1976-2020 by studing external resources.


- @Anitha & @Verna
- Note: if possible please add the reference of the resources





In [ ]:
valid_state_po_lst = []
valid_state_po_lst = US_state_info['STATE_PO'].tolist()

In [ ]:
isValid_state_po_lst = df['state_po'].isin(valid_state_po_lst)

In [ ]:
if isValid_state_po_lst.all():
  print('All the values under STATE_PO columns are validated')
else:
  print('All the values under STATE_PO columns are not validated')
  print('Invalid Values are')
  print(df[~isValid_state_po_lst])

All the values under STATE_PO columns are validated


In [ ]:
US_state_info.head()

,STATE,STATE_PO,STATE_FIPS,STATE_CENSUS_CODE,STATE_ICPSR
0,ALABAMA,AL,1,NaN,41
1,ALASKA,AK,2,NaN,81
2,ARIZONA,AZ,4,NaN,61
3,ARKANSAS,AR,5,NaN,42
4,CALIFORNIA,CA,6,NaN,71


In [ ]:
# Approach 2
set_state_po = set(zip(US_state_info['STATE'],US_state_info['STATE_PO']))

isValid_state_po = df.apply( lambda row: (row['state'],row['state_po']) in set_state_po, axis = 1  )

if isValid_state_po.all():
  print('All the values under STATE_PO columns are validated')
else:
  print('All the values under STATE_PO columns are not validated')
  print('Invalid Values are')
  print(df[~isValid_state_po])

All the values under STATE_PO columns are validated


**STATE_FIPS**: The Federal Information Processing Standard (FIPS) code for the state (4 for Arizona).
FIPS codes are unique numerical identifiers assigned to geographic areas.



*   Need to study
*   @Parth
- Need to include resource reference



In [ ]:
US_state_info.head()

,STATE,STATE_PO,STATE_FIPS,STATE_CENSUS_CODE,STATE_ICPSR
0,ALABAMA,AL,1,NaN,41
1,ALASKA,AK,2,NaN,81
2,ARIZONA,AZ,4,NaN,61
3,ARKANSAS,AR,5,NaN,42
4,CALIFORNIA,CA,6,NaN,71


In [ ]:
set_state_fips = set(zip(US_state_info['STATE'],US_state_info['STATE_FIPS']))

isValid_state_fips = df.apply( lambda row: (row['state'],row['state_fips']) in set_state_fips, axis = 1  )

if isValid_state_fips.all():
  print('All the values under STATE_FIPS columns are validated')
else:
  print('All the values under STATE_FIPS columns are not validated')
  print('Invalid Values are')
  print(df[~isValid_state_fips])

All the values under STATE_FIPS columns are validated


**STATE_CEN**: Census Code for the state (86 for Arizona).
Used by the U.S. Census Bureau for statistical purposes.

- Need to Study
- @Anitha & @Verna
- Note: if possible please add the reference of the resources





In [ ]:
US_state_info.head()

,STATE,STATE_PO,STATE_FIPS,STATE_CENSUS_CODE,STATE_ICPSR
0,ALABAMA,AL,1,NaN,41
1,ALASKA,AK,2,NaN,81
2,ARIZONA,AZ,4,NaN,61
3,ARKANSAS,AR,5,NaN,42
4,CALIFORNIA,CA,6,NaN,71


In [ ]:
set_state_ic = set(zip(US_state_info['STATE'],US_state_info['STATE_ICPSR']))

isValid_state_ic = df.apply( lambda row: (row['state'],row['state_ic']) in set_state_ic, axis = 1  )

if isValid_state_ic.all():
  print('All the values under STATE_IC columns are validated')
else:
  print('All the values under STATE_IC columns are not validated')
  print('Invalid Values are')
  print(df[~isValid_state_ic])

All the values under STATE_IC columns are validated


**STATE_IC**: Internal Census Bureau code for the state (61 for Arizona).
Similar to state_cen, it’s an internal identifier

- Need to Study
- @Anitha & @Verna
- Note: if possible please add the reference of the resources





In [ ]:
df['state_ic'].nunique()

50

In [ ]:
df['state_ic'].unique()

array([61, 71,  1, 11, 43, 82, 22,  2, 52,  3, 23, 33, 46, 34, 64, 35, 65,
       12, 66, 13, 36, 24, 14,  5, 54, 49, 67,  6, 40, 73, 56, 25, 68, 41,
       81, 42, 62, 44, 63, 21, 31, 32, 51, 45,  4, 47, 53, 72, 48, 37])

**OFFICE**: The political office contested in the election (US SENATE in this dataset).
Other values in a broader dataset might include "US HOUSE," "GOVERNOR," etc.

- Need to make sure all the values are US SENATE

In [ ]:
df['office'].unique()

array(['US SENATE'], dtype=object)

In [ ]:
if (df['office'] == 'US SENATE').all():
  print('All the values in office attribute has same value which is US SENATE')
else:
  print('Office attribute has values other then US SENATE')

All the values in office attribute has same value which is US SENATE


**DISTRICT**: The congressional district for the election.
Since U.S. Senate races are state-wide, this value is always "statewide".

- Need to make sure all the values are 'statewide'

In [ ]:
df['district'].unique()

array(['statewide'], dtype=object)

In [ ]:
if (df['district'] == 'statewide').all():
  print('All the values in district attribute has same value which is statewide')
else:
  print('district attribute has values other then statewide')

All the values in district attribute has same value which is statewide


**STAGE**: The stage of the election, here "gen" & "pre"
- Need to check unique value
- Need to study
- Categorial values
- @parth


In [ ]:
df['stage'].unique()

array(['gen', 'pre'], dtype=object)

In [ ]:
df['stage'].value_counts()

,count
stage,
gen,3616
pre,9


**SPECIAL**: Indicates if the election was a special election (due to vacancy, etc.).
False means it was a regular election.

- Binary classificaiton values
- check for other value other that True/False
- @Parth

In [ ]:
df['special'].unique()

array([False,  True])

In [ ]:
df['special'].value_counts()

,count
special,
False,3480
True,145


**CANDIDATE**: The name of the candidate running in the election.

- It has Missing values, Need to address those
- replacing missing values with 'NOT GIVEN'
- @Parth

In [ ]:
df['candidate'].nunique()

2378

In [ ]:
df['candidate'].value_counts()[:10]

,count
candidate,
OTHER,57
SCATTER,26
BLANK VOTE/SCATTERING,20
BLANK VOTE,12
CHARLES E. SCHUMER,11
OVER VOTE,9
KIRSTEN E. GILLIBRAND,8
THAD COCHRAN,7
JERRY LEVY,7


In [ ]:
df['candidate'].isnull().sum()

423

In [ ]:
df['candidate'].fillna('NOT GIVEN',inplace=True)

<ipython-input-40-c8fb86525ccf>:1: FutureWarning:

A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.



<ipython-input-40-c8fb86525ccf>:1: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



In [ ]:
df['candidate'].value_counts()[:5]

,count
candidate,
NOT GIVEN,423
OTHER,57
SCATTER,26
BLANK VOTE/SCATTERING,20
BLANK VOTE,12


**PARTY_DETAILED**: The detailed political party affiliation of the candidate.
"REPUBLICAN"
"DEMOCRAT"
"LIBERTARIAN"
"INDEPENDENT"

- It has Missing values, Need to address those
- replacing missing values with 'NOT GIVEN'
- @Parth

In [ ]:
df['party_detailed'].nunique()

172

In [ ]:
df['party_detailed'].value_counts()[:10]

,count
party_detailed,
REPUBLICAN,801
DEMOCRAT,795
LIBERTARIAN,361
INDEPENDENT,280
GREEN,87
SOCIALIST WORKERS,65
NATURAL LAW,46
CONSTITUTION,41
REFORM,36


In [ ]:
df['party_detailed'].isnull().sum()

599

In [ ]:
df['party_detailed'].fillna('NOT GIVEN', inplace=True)

<ipython-input-45-d1cf226b45ac>:1: FutureWarning:

A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.



<ipython-input-45-d1cf226b45ac>:1: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



In [ ]:
df['party_detailed'].value_counts()[:5]

,count
party_detailed,
REPUBLICAN,801
DEMOCRAT,795
NOT GIVEN,599
LIBERTARIAN,361
INDEPENDENT,280


**WRITEIN**: True if the candidate was a write-in (not officially on the ballot).
False means they were officially listed on the ballot.

- Binary classificaiton (True/False)
- Need to study and understand
- @athul
- Note: If possible please add reference of the study material

In [ ]:
df['writein'].unique()

array([False,  True])

In [ ]:
df['writein'].value_counts()

,count
writein,
False,3153
True,472


**MODE**: The method of voting recorded, always "total" in this dataset.
Could indicate specific voting methods (mail-in, early voting, etc.) in broader datasets.

- ALl the values should be 'Total'
- check if other then 'Total' value exists
- @Parth

In [ ]:
df['mode'].unique()

array(['total'], dtype=object)

In [ ]:
if (df['mode'] == 'total').all():
  print('All the values in mode attribute has same value which is total')
else:
  print('mode attribute has values other then total')

All the values in mode attribute has same value which is total


**CANDIDATEVOTES**: The total number of votes received by the candidate.

- Need to check that all the values should be less then it corressponding TOTALVOTES values. and should be positive number and greater then zero.
- @Parth

In [ ]:
df['candidatevotes'].info()

<class 'pandas.core.series.Series'>
Index: 3625 entries, 0 to 3624
Series name: candidatevotes
Non-Null Count  Dtype
--------------  -----
3625 non-null   int64
dtypes: int64(1)
memory usage: 56.6 KB


In [ ]:
df['candidatevotes'].describe()

,candidatevotes
count,3.625000e+03
mean,4.076929e+05
std,7.770793e+05
min,1.000000e+00
25%,4.792000e+03
50%,5.469900e+04
75%,4.765720e+05
max,7.864624e+06


In [ ]:
fig = px.violin(df, y='candidatevotes', title='Violin Plot of candidatevotes',orientation='v')
fig.show()

In [ ]:
fig = px.histogram(df, x='candidatevotes', title='Histogram of candidatevotes')
fig.update_layout(bargap=0.2)
fig.show()

NameError: name 'px' is not defined

**TOTALVOTES**: The total number of votes cast in the election for all candidates combined

- Need to check to identify a specific range, and number should be belongs to in between the range.
- Ideally need to cross check with population data in the given time period
- @Parth

In [ ]:
df['totalvotes'].info()

<class 'pandas.core.series.Series'>
Index: 3625 entries, 0 to 3624
Series name: totalvotes
Non-Null Count  Dtype
--------------  -----
3625 non-null   int64
dtypes: int64(1)
memory usage: 56.6 KB


In [ ]:
df['totalvotes'].describe()

,totalvotes
count,3.625000e+03
mean,2.221757e+06
std,2.113931e+06
min,1.000000e+00
25%,6.970120e+05
50%,1.576857e+06
75%,2.915281e+06
max,1.257851e+07


In [ ]:
fig = px.violin(df, y='totalvotes', title='Violin Plot of totalvotes',orientation='v')
fig.show()

In [ ]:
fig = px.histogram(df, x='totalvotes', title='Histogram of totalvotes')
fig.update_layout(bargap=0.2)
fig.show()

**UNOFFICIAL**: Indicates whether the election result is official or unofficial.
False means the results are official.

- Binary classificaiton (True/False)
- Need to study and understand
- @Parth
- Note: If possible please add reference of the study material

In [ ]:
df['unofficial'].unique()

array([False,  True])

In [ ]:
df['unofficial'].value_counts()

,count
unofficial,
False,3620
True,5


**VERSION**: A timestamp or version number for data tracking purposes.
"20210114" suggests the dataset was last updated on January 14, 2021.

*   Value should be within the timestamp range of 1976 to 2020



In [ ]:
df['version'].unique()

array([20210114])

**PARTY_SIMPLIFIED**: A simplified version of party_detailed, categorizing candidates into broader political groups:

"REPUBLICAN" → Republican Party

"DEMOCRAT" → Democratic Party

"OTHER" → Independents or minor parties

"LIBERTARIAN" → Libertarian Party

- Need to check unique value
- Need to study
- Categorial values
- @parth

In [ ]:
df['party_simplified'].info()

<class 'pandas.core.series.Series'>
Index: 3625 entries, 0 to 3624
Series name: party_simplified
Non-Null Count  Dtype 
--------------  ----- 
3625 non-null   object
dtypes: object(1)
memory usage: 56.6+ KB


In [ ]:
df['party_simplified'].unique()

array(['REPUBLICAN', 'OTHER', 'DEMOCRAT', 'LIBERTARIAN'], dtype=object)

In [ ]:
df['party_simplified'].value_counts()

,count
party_simplified,
OTHER,1656
DEMOCRAT,807
REPUBLICAN,801
LIBERTARIAN,361


In [ ]:
fig = px.histogram(df, x='party_simplified', title='Histogram of party_simplified')
fig.update_layout(bargap=0.2)
fig.show()

We have processed the data

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 3625 entries, 0 to 3624
Data columns (total 19 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   year              3625 non-null   int64 
 1   state             3625 non-null   object
 2   state_po          3625 non-null   object
 3   state_fips        3625 non-null   int64 
 4   state_cen         3625 non-null   int64 
 5   state_ic          3625 non-null   int64 
 6   office            3625 non-null   object
 7   district          3625 non-null   object
 8   stage             3625 non-null   object
 9   special           3625 non-null   bool  
 10  candidate         3625 non-null   object
 11  party_detailed    3625 non-null   object
 12  writein           3625 non-null   bool  
 13  mode              3625 non-null   object
 14  candidatevotes    3625 non-null   int64 
 15  totalvotes        3625 non-null   int64 
 16  unofficial        3625 non-null   bool  
 17  version           3

In [ ]:
df.isnull().sum()

,0
year,0
state,0
state_po,0
state_fips,0
state_cen,0
state_ic,0
office,0
district,0
stage,0
special,0


**Exporting the processed dataframe.**

In [ ]:
df.head()

,year,state,state_po,state_fips,state_cen,state_ic,office,district,stage,special,candidate,party_detailed,writein,mode,candidatevotes,totalvotes,unofficial,version,party_simplified
0,1976,ARIZONA,AZ,4,86,61,US SENATE,statewide,gen,False,SAM STEIGER,REPUBLICAN,False,total,321236,741210,False,20210114,REPUBLICAN
1,1976,ARIZONA,AZ,4,86,61,US SENATE,statewide,gen,False,WM. MATHEWS FEIGHAN,INDEPENDENT,False,total,1565,741210,False,20210114,OTHER
2,1976,ARIZONA,AZ,4,86,61,US SENATE,statewide,gen,False,DENNIS DECONCINI,DEMOCRAT,False,total,400334,741210,False,20210114,DEMOCRAT
3,1976,ARIZONA,AZ,4,86,61,US SENATE,statewide,gen,False,ALLAN NORWITZ,LIBERTARIAN,False,total,7310,741210,False,20210114,LIBERTARIAN
4,1976,ARIZONA,AZ,4,86,61,US SENATE,statewide,gen,False,BOB FIELD,INDEPENDENT,False,total,10765,741210,False,20210114,OTHER


In [ ]:
df.columns

Index(['year', 'state', 'state_po', 'state_fips', 'state_cen', 'state_ic',
       'office', 'district', 'stage', 'special', 'candidate', 'party_detailed',
       'writein', 'mode', 'candidatevotes', 'totalvotes', 'unofficial',
       'version', 'party_simplified'],
      dtype='object')

In [ ]:
# df.to_excel('Processed_1976-2020-senate.xlsx', sheet_name = 'processed',
#             columns=['year','state','stage','special','candidate',
#                         'party_detailed','writein','candidatevotes','totalvotes',
#                         'unofficial','party_simplified'],
#             index=False)

In [ ]:
import pandas as pd

In [ ]:
pdf = pd.read_excel('./Processed_1976-2020-senate.xlsx')

In [ ]:
result_1 = pdf.groupby(['year','state'])[['candidatevotes']].max()

In [ ]:
result_1.columns

Index(['candidatevotes'], dtype='object')

In [ ]:
result_1.head()

candidatevotes
year state                      
1976 ARIZONA              400334
     CALIFORNIA          3748973
     CONNECTICUT          785683
     DELAWARE             125454
     FLORIDA             1799518

In [ ]:
result_1.to_excel('Election_winner_2.xlsx', sheet_name = 'elec_win2',
            columns=['year','state','stage','special','candidate',
                        'party_detailed','writein','candidatevotes','totalvotes',
                        'unofficial','party_simplified'],
            index=False)

In [ ]:
len(result_1)

774

In [ ]:
result = pdf.groupby(['year','state'])[['candidatevotes','party_simplified']].max()

In [ ]:
len(result)

774